batch/test/test_batch.py

```diff
@@ -1777,3 +1793,417 @@ def test_region(client: BatchClient):
+    b = create_batch(client)
+    b.create_job_group(attributes={'name': 'foo', 'test': '1'})
+    b.submit()
+    job_groups = list(b.job_groups())
+    assert len(job_groups) == 1, str(job_groups)
+    jg = job_groups[0]
+    assert jg.attributes() == {'name': 'foo', 'test': '1'}, str(jg.debug_info())
+
+
+def test_job_groups_with_slow_create(client: BatchClient):
+    b = create_batch(client)
+    b.create_job_group(attributes={'name': 'foo'})
+    for _ in range(4):
+        b.create_job(DOCKER_ROOT_IMAGE, ['echo', 'a' * (900 * 1024)])
+    b.submit()
+    job_groups = list(b.job_groups())
+    assert len(job_groups) == 1, str(job_groups)
+    jobs = list(b.jobs())
+    assert len(jobs) == 4, str(jobs)
+
+
+def test_job_groups_with_slow_update(client: BatchClient):
+    b = create_batch(client)
+    jg = b.create_job_group(attributes={'name': 'foo'})
+    b.submit()
+
+    for _ in range(4):
+        jg.create_job(DOCKER_ROOT_IMAGE, ['echo', 'a' * (900 * 1024)])
+    b.submit()
+
+    status = b.status()
+    assert status['n_jobs'] == 4, str(b.debug_info())
+    assert len(list(b.job_groups())) == 1, str(b.debug_info())
+
+
+def test_more_than_one_bunch_of_job_groups_created(client: BatchClient):
+    max_bunch_size = AioBatch.MAX_BUNCH_SIZE
+    b = create_batch(client)
+    for i in range(max_bunch_size + 1):
+        b.create_job_group(attributes={'name': f'foo{i}'})
+    b.submit()
+    job_groups = list(b.job_groups())
+    assert len(job_groups) == max_bunch_size + 1, str(job_groups)
+
+
+def test_more_than_one_bunch_of_job_groups_updated(client: BatchClient):
+    max_bunch_size = AioBatch.MAX_BUNCH_SIZE
+    b = create_batch(client)
+    b.create_job_group(attributes={'name': 'foo'})
+    b.submit()
+    for i in range(max_bunch_size + 1):
+        b.create_job_group(attributes={'name': f'foo{i}'})
+    b.submit()
+    job_groups = list(b.job_groups())
+    # need to include the initial job group created
+    assert len(job_groups) == max_bunch_size + 2, str(job_groups)
+
+
+def test_job_group_cancel_after_n_failures(client: BatchClient):
+    b = create_batch(client)
+    jg = b.create_job_group(cancel_after_n_failures=1)
+    jg.create_job(DOCKER_ROOT_IMAGE, ['false'])
+    j2 = jg.create_job(DOCKER_ROOT_IMAGE, ['sleep', '300'])
+    b.submit()
+    j2_status = j2.wait()
+    jg_status = jg.wait()
+    assert j2_status['state'] == 'Cancelled', str((j2_status, jg.debug_info()))
+    assert jg_status['state'] == 'failure', str((jg_status, jg.debug_info()))
+
+
+def test_cancel_job_group(client: BatchClient):
+    b = create_batch(client)
+    jg = b.create_job_group()
+    head = jg.create_job(DOCKER_ROOT_IMAGE, ['sleep', '300'])
+    tail = jg.create_job(DOCKER_ROOT_IMAGE, ['true'], parents=[head])
+    b.submit()
+
+    head._wait_for_states('Running')
+
+    jg.cancel()
+    b_status = b.wait()
+    jg_status = jg.status()
+
+    assert b_status['state'] == 'cancelled', str(b_status)
```

Comment:
Why is the batch cancelled? It looks to me like you've only cancelled a single job group within the batch.